junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt

```diff
@@ -19,13 +19,28 @@ import org.junit.jupiter.api.function.ThrowingSupplier
 import java.time.Duration
 import java.util.function.Supplier
 import java.util.stream.Stream
+import kotlin.contracts.InvocationKind.AT_MOST_ONCE
+import kotlin.contracts.InvocationKind.EXACTLY_ONCE
+import kotlin.contracts.contract
 
 /**
  * @see Assertions.fail
  */
 fun fail(message: String?, throwable: Throwable? = null): Nothing =
     Assertions.fail<Nothing>(message, throwable)
 
+/**
+ * @see Assertions.fail
+ */
+@JvmName("fail_nonNullableLambda")
+fun fail(message: () -> String): Nothing {
+    contract {
+        callsInPlace(message)
+    }
+
+    return Assertions.fail<Nothing>(message)
```

Comment:
This return looks redundant, but might be necessary because of Java interop?

Reply:
Yes. Since kotlin compiler doesn't know that `Assertions.fail` never returns, it requires us to return in this method